helper/keysutil/encrypted_key_storage.go

```diff
@@ -0,0 +1,243 @@
+		return nil, err
+	}
+
+	return &EncryptedKeyStorage{
+		policy: config.Policy,
+		s:      config.Storage,
+		prefix: config.Prefix,
+		lru:    cache,
+	}, nil
+}
+
+// EncryptedKeyStorage implements the logical.Storage interface and ensures the
+// storage paths are encrypted in the underlying storage.
+type EncryptedKeyStorage struct {
+	policy *Policy
+	s      logical.Storage
+	lru    *lru.TwoQueueCache
+
+	prefix string
+}
+
+// List implements the logical.Storage List method, and decrypts all the items
+// in a path prefix. This can only operate on full folder structures so the
+// prefix should end in a "/".
+func (s *EncryptedKeyStorage) List(ctx context.Context, prefix string) ([]string, error) {
+	encPrefix, err := s.encryptPath(prefix)
+	if err != nil {
+		return nil, err
+	}
+
+	keys, err := s.s.List(ctx, encPrefix+"/")
+	if err != nil {
+		return keys, err
+	}
+
+	decryptedKeys := make([]string, len(keys))
+
+	// The context for the decryption operations will be the object's prefix
+	// joined with the provided prefix. Join cleans the path ensuring there
+	// isn't a trailing "/".
+	context := []byte(paths.Join(s.prefix, prefix))
+
+	for i, k := range keys {
+		raw, ok := s.lru.Get(k)
+		if ok {
+			// cahce HIT, we can bail early and skip the decode & decrypt operations.
+			decryptedKeys[i] = raw.(string)
+			continue
+		}
+
+		// If a folder is included in the keys it will have a trailing "/".
+		// We need to remove this before decoding/decrypting and add it back
+		// later.
+		appendSlash := strings.HasSuffix(k, "/")
+		if appendSlash {
+			k = strings.TrimSuffix(k, "/")
+		}
+
+		decoded := base58.Decode(k)
+		if len(decoded) == 0 {
+			return nil, errors.New("Could not decode key")
+		}
+
+		// Decrypt the data with the object's key policy.
+		encodedPlaintext, err := s.policy.Decrypt(context, nil, string(decoded[:]))
+		if err != nil {
+			return nil, err
+		}
+
+		// The plaintext is still base64 encoded, decode it.
+		decoded, err = base64.StdEncoding.DecodeString(encodedPlaintext)
+		if err != nil {
+			return nil, err
+		}
+
+		plaintext := string(decoded[:])
+
+		// Add the slash back to the plaintext value
+		if appendSlash {
+			plaintext += "/"
+			k += "/"
+		}
+
+		// We want to store the unencoded version of the key in the cache.
+		// This will make it more performent when it's a HIT.
+		s.lru.Add(k, plaintext)
+
+		decryptedKeys[i] = plaintext
+	}
+
+	return decryptedKeys, nil
+}
+
+// Get implements the logical.Storage Get method.
+func (s *EncryptedKeyStorage) Get(ctx context.Context, path string) (*logical.StorageEntry, error) {
+	encPath, err := s.encryptPath(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.s.Get(ctx, encPath)
+}
+
+// Put implements the logical.Storage Put method.
+func (s *EncryptedKeyStorage) Put(ctx context.Context, entry *logical.StorageEntry) error {
+	encPath, err := s.encryptPath(entry.Key)
+	if err != nil {
+		return err
+	}
+	e := &logical.StorageEntry{}
+	*e = *entry
```

Comment:
Is there an advantage of doing this?

Reply:
Yeah it prevents the entry that was passed in from being chaghed to the encrypted path